artmaker/src/main/java/io/artmaker/composables/ArtMakerDrawScreen.kt

```diff
@@ -147,39 +161,43 @@ internal fun ArtMakerDrawScreen(
                     drawLayer(graphicsLayer)
                 }
             }
-            .pointerInput(Unit) {
-                detectTapGestures(
-                    onTap = { offset ->
-                        onDrawEvent(DrawEvent.AddNewShape(offset))
-                    },
-                )
-            }
-            .pointerInput(isFullScreenMode) {
-                detectDragGestures(
-                    onDragStart = { offset ->
+            .pointerInteropFilter { event ->
+                val offset = Offset(event.x, event.y)
+                when (event.action) {
+                    MotionEvent.ACTION_DOWN -> {
+                        getDialogType(context, event, state.shouldUseStylusOnly)?.let { type ->
+                            shouldShowStylusDialog = true
+                            stylusDialogType = type
+                        }
+
+                        if (!event.validateEvent(context, state.shouldUseStylusOnly)) return@pointerInteropFilter false
+
                         onDrawEvent(DrawEvent.AddNewShape(offset))
-                    },
-                    onDragCancel = {
+                    }
+
+                    MotionEvent.ACTION_MOVE -> {
+                        val clampedOffset =
+                            Offset(x = offset.x, y = clamp(offset.y, 0f, maxDrawingHeight))
+                        onDrawEvent(DrawEvent.UpdateCurrentShape(clampedOffset))
+                    }
+
+                    MotionEvent.ACTION_CANCEL -> {
                         onDrawEvent(DrawEvent.UndoLastShapePoint)
```

Comment:
From the docs:

    /**
     * Constant for {@link #getActionMasked}: The current gesture has been aborted.
     * You will not receive any more points in it.  You should treat this as
     * an up event, but not perform any action that you normally would.
     */
     
     The last line says that `You should treat this as an up event, but not perform any action that you normally would.` I admit that it is confusing. Here we can leave it because we need to ensure that any of the usual event, which is drawing in our case, has not been done. If done, undo it.